File: internal/model/strength_forecast.go
package model

import "time"

type StrengthForecast struct {
	ID                    uint              `gorm:"primaryKey" json:"id"`
	PourSectionID         uint              `gorm:"not null;index" json:"pour_section_id"`
	PourSection           PourSection       `gorm:"foreignKey:PourSectionID" json:"pour_section,omitempty"`
	TemperatureSeriesID   uint              `gorm:"not null;index" json:"temperature_series_id"`
	TemperatureSeries     TemperatureSeries `gorm:"foreignKey:TemperatureSeriesID" json:"temperature_series,omitempty"`
	MixDesignVersion      int               `gorm:"not null" json:"mix_design_version"`
	FormulaVersion        string            `gorm:"size:48;not null;index;uniqueIndex:idx_forecast_input" json:"formula_version"`
	InputHash             string            `gorm:"size:64;not null;index;uniqueIndex:idx_forecast_input" json:"input_hash"`
	InputSnapshot         string            `gorm:"type:text;not null" json:"input_snapshot"`
	MaturityDegreeHours   float64           `gorm:"not null" json:"maturity_degree_hours"`
	PredictedStrengthMPA  float64           `gorm:"not null" json:"predicted_strength_mpa"`
	ThresholdETA          *time.Time        `json:"threshold_eta,omitempty"`
	ConfidenceLevel       string            `gorm:"size:16;not null;index" json:"confidence_level"`
	ForecastState         string            `gorm:"size:24;not null;index" json:"forecast_state"`
	Explanation           string            `gorm:"type:text;not null" json:"explanation"`
	CalculatedBy          uint              `gorm:"not null;index" json:"calculated_by"`
	CalculatedByName      string            `gorm:"size:120;not null" json:"calculated_by_name"`
	CalculatedAt          time.Time         `gorm:"not null" json:"calculated_at"`
	ReviewedBy            *uint             `json:"reviewed_by,omitempty"`
	ReviewedByName        string            `gorm:"size:120" json:"reviewed_by_name,omitempty"`
	ReviewedAt            *time.Time        `json:"reviewed_at,omitempty"`
	ConfirmedBy           *uint             `json:"confirmed_by,omitempty"`
	ConfirmedAt           *time.Time        `json:"confirmed_at,omitempty"`
	IdempotencyKey        string            `gorm:"size:100;not null;uniqueIndex:idx_actor_idempotency" json:"idempotency_key"`
	IdempotencyActorID    uint              `gorm:"not null;uniqueIndex:idx_actor_idempotency" json:"idempotency_actor_id"`
	DurationMilliseconds  int64             `gorm:"not null" json:"duration_milliseconds"`
	FailureReason         string            `gorm:"type:text" json:"failure_reason,omitempty"`
	DeterminismReplayPass *bool             `json:"determinism_replay_passed,omitempty"`
	CreatedAt             time.Time         `json:"created_at"`
	UpdatedAt             time.Time         `json:"updated_at"`
}

File: frontend/src/types/strength-forecast.ts
import type { ConfidenceLevel } from './enums/confidence-level'

export type ForecastState = 'queued' | 'calculating' | 'completed' | 'failed' | 'reviewed' | 'confirmed' | 'voided'

export interface MaturityStep {
  index: number
  from: string
  to: string
  average_temperature_c: number
  duration_hours: number
  contribution_degree_hours: number
  running_maturity_degree_hours: number
  below_datum_clamped: boolean
}

export interface ForecastExplanation {
  formula_version: string
  formula: string
  below_datum_rule: string
  maturity: { degree_hours: number; duration_hours: number; steps: MaturityStep[] }
  interpolation: Record<string, unknown>
  data_coverage_percent: number
  quality_warnings: string[]
  boundary_note: string
  target_strength_mpa: number
  recent_maturity_rate_per_hour: number
}

export interface StrengthForecast {
  id: number
  pour_section_id: number
  section_code: string
  temperature_series_id: number
  sensor_code: string
  mix_design_version: number
  formula_version: string
  input_hash: string
  input_snapshot: Record<string, unknown>
  maturity_degree_hours: number
  predicted_strength_mpa: number
  threshold_eta?: string
  confidence_level: ConfidenceLevel
  forecast_state: ForecastState
  explanation: ForecastExplanation
  calculated_by: number
  calculated_by_name: string
  calculated_at: string
  reviewed_by?: number
  reviewed_by_name?: string
  confirmed_by?: number
  idempotency_key: string
  duration_milliseconds: number
  failure_reason?: string
  determinism_replay_passed?: boolean
  created_at: string
}
